README.md
# Random Mixed Field Model
Demo code of our AAAI 2016 paper "Random Mixed Field Model for Mixed-Attribute Data Restoration".

%% Code authors: Qiang Li

%% Release time: Jun. 3rd, 2018

%% Current version: RMF_code_v1

0. Add path.

First of all, run 'AddPath.m' to add paths of required packages.

1. Run the Main code.

There are three demos, i.e.,

'DemoClassifyMissNoisyData.m' gives classification comparison.

'DemoInferMixedNet.m' presents variational inference for data denoising.
In this case, the RMF model parameters are user-defined and fixed.

It also gives the MSE plots and graph visualization.
Note there maybe cases when denoising cont/disc nodes is ineffective.
This is possibly due to a compromise between continuous and discrete nodes.

'DemoLearnStruct.m' presents structure and parameter learning.
In this demo, the RMF model is learned using Jason Lee's code.

2. Datasets and Corrupted data preparation.

The experiment is conducted on several UCI datasets.

https://archive.ics.uci.edu/ml/datasets.html

Run 'GenMissNoisyData.m' to get the corrupted data.

3. Dependencies.

UGM at http://www.di.ens.fr/~mschmidt/Software/UGM_2009.zip

TFOCS at http://tfocs.stanford.edu

MGM at http://www-bcf.usc.edu/~lee715/syntheticExp/syntheticExp.zip

%% Reference noticement:

If you have used the code, please cite the following paper:

[1] Random Mixed Field Model for Mixed-Attribute Data Restoration

Qiang Li, Wei Bian, Richard Yi Da Xu, Jane You and Dacheng Tao

AAAI Conference on Artificial Intelligence (AAAI), Feb. 2016, pp. 1244--1250.

%% Supporting information:

If any questions and comments, feel free to send your email to

Qiang Li (leetsiang.cloud@gmail.com)


RegEM/CHANGES
26-Aug-14   * Added an option to use other norms than the 2-norm for the error 
	      to be minimized in K-fold cross-validation.

15-Jul-14   * Changed default stagnation tolerance to 1e-2.
	    * Added option to scale variables abitrarily, with given vector 
	      of scale factors (rather than always standardizing variables, 
	      which is inappropriate if the relative errors on the variables 
	      differ greatly across the dataset).

14-Aug-13   * Implemented an option to choose a TTLS truncation parameter by 
	      K-fold cross-validation (added new functions KCV_TTLS and 
	      KCVINDICES to do so).

15-Jul-13   * Made code more modular, with new functions MISSINGNESS_PATTERNS 
	      and RESCALE. Updated syntax of several Matlab commands to be more 
	      in line with newer Matlab versions.

12-Nov-12   * Removed display option (always display diagnostics now)
            * In each EM iteration, regressions are now only computed for 
              each unique patterns of missigness in the data matrix, rather 
              than for each row (using code from Julien Emile-Geay). This 
	      speeds up the computations considerably if there are only a few 
	      patterns of missigness (e.g., a staircase pattern of missigness 
	      with broad 'steps')
            * Cleaned up the driver module regem.m to remove some redundancy in 
              the code

09-Jun-12   * Added some comments on the computation of residual 
              covariances to PTTLS.

18-Feb-12   * Updated PEIGS so that it calls Matlab's full eigendecomposition
              routine EIG unless the number of eigenpairs to be computed 
              (rmax) is much less than the matrix dimension (only if rmax 
              is much less than the matrix dimension is EIGS called, which 
              then is faster than EIG).

            * Added an option to select truncation parameters in
              TTLS by a variety of (information-theoretic) selection
              criteria, such as minimizers of the Akaike's
              Information Criterion or Schwarz's and Rissanen's Mean
              Description Length.

20-Oct-10   * Fixed NANSTD. (It previously calculated an incorrect
              estimate of the standard deviation.) (Error pointed
              out by Ross Tulloch.)

14-Oct-10   * Fixed syntax in REGEM line where it is checked whether a
              TTLS truncation parameter is given. (Error pointed out
              by Gavin Schmidt.)

13-Aug-09   * Removed display option in call of EIGS in PEIGS.  The
	          option created an error for a 1x1 input matrix (a bug
              in Matlab's EIGS).

03-Jul-07   * Corrected a problem in PEIGS with recent Matlab
              implementations, in which eigenvalues were not
              necessarily returned in descending order.   

15-Mar-07   * Changed screen output of effective number of variables
              (peff) in REGEM for TTLS option. REGEM was erroneously
              writing out an effective number of degrees of
              freedom, rather than an effective number of variables
              (=truncation parameter). However, this was of no
              consequence for the performance of the algorithm. 

              Note that this definition of the effective number of
              variables is not precise for TTLS. A better definition
              would be based on the filter factor formulation of
              Fierro et al. (1997), but obtaining these would
              involve additional computation, which seems
              unnecessary for this merely diagnostic output. 

01-Jul-02   * Users reported problems due to changes in the calling 
              sequence of EIGS in a beta version of an upcoming Matlab 
              release. 

08-Feb-02   * Minor changes in PEIGS and GCVRIDGE to improve downward 
              compatibility with Matlab 5. Switched to multiple ridge 
              regressions as default regularized regression method.

02-Jan-01   * Adaptation to Matlab 6 (minor changes in PEIGS). REGEM now 
              has an additional optional parameter OPTIONS.minvarfrac, 
              from which an upper bound on the regularization parameter
              can be constructed. 

04-Apr-00   * CovRes in REGEM is now a full matrix, no longer a
              sparse matrix. Not allocating memory at initialization
              of the sparse matrix had significantly slowed down the
              algorithm.

22-Mar-00   * All variables in a dataset are now scaled at the
              beginning of an EM iteration and regularization in
              standard form is performed. Before, in each individual 
              record only the variables with available values were 
              scaled. Scaling all variables slightly changes the
              objective of generalized cross-validation: instead of
              estimating the regularization parameter for which the
              expected *ms error* of the imputed values is minimum,
              GCV now estimates the regularization parameter for
              which the expected *relative ms error* of the imputed
              values is minimum. REGEM with MRIDGE might thus produce
              slightly different results; however, REGEM with IRIDGE
              should produce the same results as before.

FuncLoadData.m
function [ Utr, Vtr, Uts, Vts, L, lambda ] = FuncLoadData( foldName, splitRatio )
% Function to load UCI dataset
%
% Note: Treat label as a disc node which is in the end of matrix.

% Default params of UCI dataset
[~,name,~] = fileparts( foldName);

removFlag = 0; % whether to remove duplicate instances

switch name
    case 'abalone'
        splitFlag = 0; scaleFlag = 0;
        contAttr = [2:8]; % continuous-valued attributes index
        discAttr = 1:9; discAttr(contAttr) = []; % discrete-valued attributes index
        L = [3 3]'; % levels in each categorical variable & label
        lambda = 0.5; % large lambda for small # training instances
    case 'adult'
        splitFlag = 0; scaleFlag = 1;
        contAttr = [1 3 5 11 12 13];
        discAttr = 1:15; discAttr(contAttr) = [];
        L = [8 16 7 14 6 5 2 41 2]';
        lambda = 0.0526;
    case 'census-income'
        removFlag = 1; % REMOVE duplicates
        splitFlag = 0; scaleFlag = 1; % NOTE: as suggested, remove 25-th attr (instance weight)
        contAttr = [1 6 17 18 19 31 40]; % 7 cont
        discAttr = 1:42; discAttr( [contAttr,25] ) = []; % 33 disc & 1 Label
        L = [9,52,47,17,3,7,24,15,5,10,2,3,6,8,6,6,50,38,8,9,8,9,3,3,5,42,42,42,5,3,3,3,2,2]';
        lambda = 0.05;
    case 'cmc'
        splitFlag = 1; scaleFlag = 1;
        contAttr = [1 4];
        discAttr = 1:10; discAttr(contAttr) = [];
        L = [4 4 2 2 4 4 2 3]';
        lambda = 0.0312;
    case 'crx'
        splitFlag = 1; scaleFlag = 1;
        contAttr = [2 3 8 11 14 15];
        discAttr = 1:16; discAttr(contAttr) = [];
        L = [2 4 3 14 9 2 2 2 3 2]';
        lambda = 0.1250;
    case 'australian'
        delimiter = {' '};
        splitFlag = 1; scaleFlag = 1;
        contAttr = [2 3 7 10 13 14];
        discAttr = 1:15; discAttr(contAttr) = [];
        L = [2,3,14,9,2,2,2,3,2]';
        lambda = 0.0312;
    case 'german'
        delimiter = {' '};
        splitFlag = 1; scaleFlag = 1;
        contAttr = [2 5 8 11 13 16 18];
        discAttr = 1:21; discAttr(contAttr) = [];
        L = [4,5,11,5,5,5,3,4,3,3,4,2,2,2]';
        lambda = 0.0884;
    case 'allbp'
        delimiter = {':',',','.|'};
        splitFlag = 0; scaleFlag = 1;
        contAttr = [1 18 20 22 24 26 28];
        discAttr = 1:30; discAttr(contAttr) = [];
        L = [2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,6,3]';
        lambda = 0.0005;
    case 'toy' % synthetic data
        load([foldName,'.mat']);
        nSamples = round( size(U, 2) * splitRatio );
        trainIdx = 1:nSamples;
        testIdx = 1:size(U, 2); testIdx(trainIdx) = [];
        Utr = U(:, trainIdx); Vtr = V(:, trainIdx); % Train set
        Uts = U(:, testIdx); Vts = V(:, testIdx); % Test set
        L = max(V,[],2); % Cardinalities (including label)
        lambda = 0.05;
        return;
    otherwise
        error('Unexpected dataset name. No data loaded.')
end

if splitFlag
    load ([foldName,'data.mat']); fileData = fileData';
    
    % Split into train/test set.
    nSamples = round( size(fileData, 2) * splitRatio ); % calculate the number of observed entries
    
%     rPerm = randperm( size(fileData, 2) ); % random split
%     trainIdx = sort( rPerm(1:nSamples) );
    
    trainIdx = 1:nSamples; % direct split
    
    testIdx = 1:size(fileData, 2); testIdx(trainIdx) = [];
    
    % Train set
    Utr = fileData(contAttr, trainIdx);
    Vtr = fileData(discAttr, trainIdx);
    
    % Test set
    Uts = fileData(contAttr, testIdx);
    Vts = fileData(discAttr, testIdx);
    clear fileData;
else
    % Train set
    load ([foldName,'data.mat']); fileData = fileData';
    
    Utr = fileData(contAttr,:);
    Vtr = fileData(discAttr,:);
    clear fileData;
    
    % Test set
    load ([foldName,'test.mat']); fileData = fileData';
    
    Uts = fileData(contAttr,:);
    Vts = fileData(discAttr,:);
    clear fileData;

end

% Remove duplicate samples
if removFlag
    [Utr,Vtr] = FuncRemDup(Utr,Vtr);
    [Uts,Vts] = FuncRemDup(Uts,Vts);
end

if scaleFlag % scale cont attributes to [0,1]
    % Handle missing values
    nanIdx = isnan(Utr(:));
    if sum(nanIdx); Utr(nanIdx) = 0; end
    
    % Train set
    trainParam.Min = min(Utr,[],2);
    trainParam.Max = max(Utr,[],2);
    trainParam.Rng = max(Utr,[],2) - min(Utr,[],2);
    trainParam.Rng( trainParam.Rng == 0 ) = 1; % avoid 0 dividend.
    
    Utr = GetScaledContAttr(Utr,trainParam); 
    
    % Test set
    Uts = GetScaledContAttr(Uts,trainParam);
end

function [X,Y] = FuncRemDup(U,V)
% Function to remove duplicate columns (i.e., instances)
X = U; Y = V;

nanElemVal = -1000;
U( isnan(U) ) = nanElemVal; % any const.
V( isnan(V) ) = nanElemVal; % any const.

[~,idx_uni] = unique([U;V]','rows');
idx_uni = sort(idx_uni);

X = X(:,idx_uni);
Y = Y(:,idx_uni);
